tests/test_grpc_functions.py

```diff
@@ -0,0 +1,248 @@
+            'GRPC_PORT': str(port),
+            'FUNCTION_URI': 'file://%s/tests/functions/upper.py?handler=handle' % os.getcwd()
+        }
+
+        self.process = subprocess.Popen(self.command,
+                                        cwd=self.workingdir,
+                                        shell=True,
+                                        env=env,
+                                        preexec_fn=os.setsid,
+                                        )
+        time.sleep(0.5)
+
+        channel = grpc.insecure_channel('localhost:%s' % port)
+        self.stub = function.MessageFunctionStub(channel)
+
+        def generate_messages():
+            headers = {
+                'Content-Type': message.Message.HeaderValue(values=['text/plain']),
+                'correlationId': message.Message.HeaderValue(values=[str(uuid.uuid4())])
+            }
+
+            messages = [
+                message.Message(payload="hello", headers=headers),
+                message.Message(payload="world", headers=headers),
+                message.Message(payload="foo", headers=headers),
+                message.Message(payload="bar", headers=headers),
+            ]
+            for msg in messages:
+                yield msg
+
+        responses = self.stub.Call(generate_messages())
+        expected = ['HELLO', 'WORLD', 'FOO', 'BAR']
+
+        for response in responses:
+            self.assertTrue(response.payload in expected)
+            expected.remove(response.payload)
+
+        self.assertEquals(0, len(expected))
+
+    def test_concat(self):
+        port = find_free_port()
+        env = {
+            'PYTHONPATH': '%s/tests/functions:$PYTHONPATH' % os.getcwd(),
+            'GRPC_PORT': str(port),
+            'FUNCTION_URI': 'file://%s/tests/functions/concat.py?handler=concat' % os.getcwd()
+        }
+
+        self.process = subprocess.Popen(self.command,
+                                        cwd=self.workingdir,
+                                        shell=True,
+                                        env=env,
+                                        preexec_fn=os.setsid,
+                                        )
+        time.sleep(0.5)
+
+        channel = grpc.insecure_channel('localhost:%s' % port)
+        self.stub = function.MessageFunctionStub(channel)
+
+        def generate_messages():
+            headers = {
+                'Content-Type': message.Message.HeaderValue(values=['application/json']),
+                'correlationId': message.Message.HeaderValue(values=[str(uuid.uuid4())])
+            }
+
+            messages = [
+                message.Message(payload='{"foo":"bar","hello":"world"}', headers=headers),
+            ]
+            for msg in messages:
+                yield msg
+
+        responses = self.stub.Call(generate_messages())
+
+        for response in responses:
+            self.assertEquals('{"result": "foobarhelloworld"}', response.payload)
```

Comment:
Is dict enumeration necessarily deterministic? If not, this test will flake.

Reply:
good point.  I can add a sort in the function

Reply:
done